OpenRA.Mods.Common/Activities/Enter.cs

```diff
@@ -127,8 +128,19 @@ public override bool Tick(Actor self)
 				{
 					// Check that we reached the requested position
 					var targetPos = target.Positions.ClosestToIgnoringPath(self.CenterPosition);
-					if (!IsCanceling && self.CenterPosition == targetPos && target.Type == TargetType.Actor)
+					if (!IsCanceling && target.Type == TargetType.Actor)
+					{
+						if (self.CenterPosition != targetPos && !hasMadeSecondApproach)
+						{
+							// We got close to the target, but not next to it, e.g. maybe the target moved slightly out of range.
+							// Since the target is likely close, approach it a second time.
+							hasMadeSecondApproach = true;
```

Comment:
I think it's not a good to assume that the transport will stop. A single pass flag does not seem correct

Reply:
Indeed, we're limited by what dervied classes will do here when `TryStartEnter` is called. The `RideTransport` activity will reserve the transport, which will lock it in place. But other derived classes do not do such an action when `TryStartEnter` is called. Though to be fair some of them are operating on targets that are usually immobile, e.g. buildings.

I ruled out retrying forever here, as it is possible to recreate the original bug with naval transports, e.g. as below:
![image](https://github.com/OpenRA/OpenRA/assets/3399086/1651c5a2-eb07-4e36-967f-c6b3786d5d38)

If you end up in this scenario, retrying forever is a big problem. The transport is locked in place and won't respond to commands as it is reserved. But the infantry cannot path to the transport - as that would require it being able to step into the water. So these units would be stuck in place if we did an endless retry. In this situation the current behaviour of cancelling the activity seems better than deadlocking the units.

So if one attempt can sometimes fail, and we don't want to restart forever, the next natural step was to try twice, which seems sufficient to fix the original bug, whilst still allowing the activity to cancel itself if the 2nd attempt fails. I figure that is useful for both the navel transport issue - where it will otherwise be stuck forever, but also useful for some weird edge case where you somehow keep getting next to the enter target, yet it has always moved by the time you try and enter it.

So I think the two attempt strategy actually makes a lot of sense here. For the `RideTransport` case when the target will be locked from moving, it neatly fixes the needing to take an extra step to complete the enter scenario. For other classes it may not be a guarentee. But limiting the retry count to 2 prevents any badness, whilst still possibly improving the number of times it will work. If the first path got you close but then something fudged it at the last moment, having a second go knowing the target is likely still close seems like a decent thing to try.

Anyway that was a long braindump of how I got to this policy as a solution - but I'm open to other options if you can think of one.

Reply:
Hmm, so the issue is that `RideTransport` assumes that if it returns true in TryStartEnter that it will either successfully enter or be cancelled. That seems a bit wrong the the unit is locked up even during the cancellation process

I think the double retry approach is more a hack than a proper solution, and should be labelled as such

Reply:
Added a `// HACK ...`